mark.c

```diff
@@ -864,40 +895,45 @@ GC_INNER mse * GC_mark_from(mse *mark_stack_top, mse *mark_stack,
           if (cheri_gettag(limit) == 0) {
             limit -= ALIGNMENT;
           } else {
-            deferred = *(word *)limit;
+            deferred = *(void **)limit;
             FIXUP_POINTER(deferred);
             limit -= ALIGNMENT;
             if (deferred >= (word)least_ha && deferred < (word)greatest_ha)
               break;
           }
           if ((word)current_p > (word)limit) goto next_object;
         }
-#       endif   // defined(__CHERI_PURE_CAPABILITY__)
-#     endif
 
-      while ((word)current_p <= (word)limit) {
-        /* Empirically, unrolling this loop doesn't help a lot. */
-        /* Since PUSH_CONTENTS expands to a lot of code,        */
-        /* we don't.                                            */
-        current = *(word *)current_p;
-        FIXUP_POINTER(current);
-        PREFETCH(current_p + PREF_DIST*CACHE_LINE_SIZE);
-        if (current >= (word)least_ha && current < (word)greatest_ha) {
-          /* Prefetch the contents of the object we just pushed.  It's  */
-          /* likely we will need them soon.                             */
-          PREFETCH((ptr_t)current);
-#         ifdef ENABLE_TRACE
-            if (GC_trace_addr == current_p) {
-              GC_log_printf("GC #%u: considering(1) %p -> %p\n",
-                            (unsigned)GC_gc_no, (void *)current_p,
-                            (void *)current);
-            }
-#         endif /* ENABLE_TRACE */
-          PUSH_CONTENTS((ptr_t)current, mark_stack_top,
-                        mark_stack_limit, current_p);
+
+        while ((word)current_p <= (word)limit) {
+          /* Empirically, unrolling this loop doesn't help a lot. */
+          /* Since PUSH_CONTENTS expands to a lot of code,        */
+          /* we don't.                                            */
+          current = *(void **)current_p;
+          if (cheri_gettag(current) == 0) {
+            current_p += ALIGNMENT;
+            continue;
+          }
```

Comment:
Ah, I was hunting for something like this, and didn't spot it!

Reply:
> Aside from `word` vs `void**`, I think this is the only change for `__CHERI_PURE_CAPABILITY__` in this `while` loop. Is that correct? (It's hard to see precisely in the diff.)
> 
> If so, could the loop be kept shared, with a quick `#ifdef` for this special case? It might make it easier to compare with and without CHERI changes.

Along with a `continue`  if the tag is invalid to prevent it pushing onto the mark stack. Changed in [c1adac6](https://github.com/capablevms/bdwgc/pull/13/commits/c1adac6865660965561975acc3587513d3299e8e)